runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp2Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+   TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp2Reg);
+   TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
+   generateCompareInstruction(cg, node, tmp1Reg, tmp2Reg);
+
+   // If either object is known to be of primitive component type,
+   // we are done: since both of them have to be of equal class.
+   if (node->isArrayChkPrimitiveArray1() || node->isArrayChkPrimitiveArray2())
+      {
+      if (NULL == snippetLabel)
+         {
+         snippetLabel = generateLabelSymbol(cg);
+         gcPoint = generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_NE);
+         snippet = new (cg->trHeapMemory()) TR::ARM64HelperCallSnippet(cg, node, snippetLabel, node->getSymbolReference(), doneLabel);
+         cg->addSnippet(snippet);
+         }
+      else
+         generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_NE);
+      }
+   // We have to take care of the un-equal class situation: both of them must be of reference array
+   else
+      {
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ);
+
+      // Object1 must be of reference component type, otherwise throw exception
+      if (!node->isArrayChkReferenceArray1())
+         {
+
+         // Loading the Class Pointer -> classDepthAndFlags
+#ifdef OMR_GC_COMPRESSED_POINTERS
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+         generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+         TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
+         generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+               new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+         // We need to perform a X & OBJECT_HEADER_SHAPE_MASK
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andimmw, node, tmp1Reg, tmp1Reg, OBJECT_HEADER_SHAPE_MASK << J9AccClassRAMShapeShift);
```

Comment:
The instruction encoding of `andimmw` is tricky.  You need to look at `logicBinaryEvaluator()` in OMR.

Reply:
`logicImmediateHelper()` is now available from the outside of OMR BinaryEvaluator.cpp.
You can encode the immediate value for `andimmw` using it.